arangod/Aql/SortedRowsStorageBackendMemory.cpp

```diff
@@ -0,0 +1,218 @@
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Jan Steemann
+////////////////////////////////////////////////////////////////////////////////
+
+#include "SortedRowsStorageBackendMemory.h"
+
+#include "Aql/AqlItemBlockInputRange.h"
+#include "Aql/ExecutionState.h"
+#include "Aql/InputAqlItemRow.h"
+#include "Aql/OutputAqlItemRow.h"
+#include "Aql/SortExecutor.h"
+#include "Aql/SortRegister.h"
+#include "Basics/Exceptions.h"
+#include "Basics/ResourceUsage.h"
+#include "Basics/debugging.h"
+
+#include <algorithm>
+
+namespace {
+using namespace arangodb;
+using namespace arangodb::aql;
+
+// custom AqlValue-aware comparator for sorting
+class OurLessThan {
+ public:
+  OurLessThan(velocypack::Options const* options,
+              std::vector<SharedAqlItemBlockPtr> const& input,
+              std::vector<SortRegister> const& sortRegisters) noexcept
+      : _vpackOptions(options), _input(input), _sortRegisters(sortRegisters) {}
+
+  bool operator()(AqlItemMatrix::RowIndex const& a,
+                  AqlItemMatrix::RowIndex const& b) const {
+    auto const& left = _input[a.first].get();
+    auto const& right = _input[b.first].get();
+    for (auto const& reg : _sortRegisters) {
+      AqlValue const& lhs = left->getValueReference(a.second, reg.reg);
+      AqlValue const& rhs = right->getValueReference(b.second, reg.reg);
+      int const cmp = AqlValue::Compare(_vpackOptions, lhs, rhs, true);
+
+      if (cmp < 0) {
+        return reg.asc;
+      } else if (cmp > 0) {
+        return !reg.asc;
+      }
+    }
+
+    return false;
+  }
+
+ private:
+  velocypack::Options const* _vpackOptions;
+  std::vector<SharedAqlItemBlockPtr> const& _input;
+  std::vector<SortRegister> const& _sortRegisters;
+};  // OurLessThan
+
+}  // namespace
+
+namespace arangodb::aql {
+
+SortedRowsStorageBackendMemory::SortedRowsStorageBackendMemory(
+    SortExecutorInfos& infos)
+    : _infos(infos),
+      _returnNext(0),
+      _memoryUsageForInputBlocks(0),
+      _sealed(false) {}
+
+SortedRowsStorageBackendMemory::~SortedRowsStorageBackendMemory() {
+  _infos.getResourceMonitor().decreaseMemoryUsage(currentMemoryUsage());
+}
+
+ExecutorState SortedRowsStorageBackendMemory::consumeInputRange(
+    AqlItemBlockInputRange& inputRange) {
+  TRI_ASSERT(!_sealed);
+
+  ExecutorState state = ExecutorState::HASMORE;
+
+  auto inputBlock = inputRange.getBlock();
+  if (inputBlock != nullptr) {
+    _inputBlocks.emplace_back(inputBlock);
+    _memoryUsageForInputBlocks += inputBlock->getMemoryUsage();
+  }
+  size_t numDataRows = inputRange.countDataRows();
+
+  ResourceUsageScope guard(_infos.getResourceMonitor());
+
+  if (_rowIndexes.capacity() < _rowIndexes.size() + numDataRows) {
+    size_t newCapacity = std::max(_rowIndexes.capacity() * 2, numDataRows);
+    while (newCapacity < _rowIndexes.size() + numDataRows) {
+      newCapacity *= 2;
+    }
+
+    // may throw
+    guard.increase((newCapacity - _rowIndexes.capacity()) *
```

Comment:
This is a good question for which may be an issue that I don't currently see, but, even though the function that increases the memory usage may throw, it is said in its description that it doesn't bogus any updates aside from the global counter that would store the memory usage, i. e.:
```
/// this function may modify up to 3 atomic variables:
/// - the current local memory usage
/// - the peak local memory usage
/// - the global memory usage
/// the order in which we update these atomic variables is not important,
/// as long as everything is eventually consistent.
/// in case this function triggers a "resource limit exceeded" error,
/// the only thing that can have happened is the update of the current local
/// memory usage, which this function will roll back again.
/// as this function only adds and subtracts values from the memory usage
/// counters, it  does not lead to any lost updates due to data races with
/// other threads.
/// the peak memory usage value is updated with a CAS operation, so again
/// there will no be lost updates.
```
so resource monitor's memory usage increase does not mess up other data that should be consistent. 
We didn't deal with the possible throwing and some other parts of the Aql code have not too, like AqlCursor.

Reply:
Do you think I should do something when throws then?

Reply:
I think you are right. Let's not do anything. If this throws, the query will probably be aborted anyway. And even if not, the fact that `_memoryUsageForInputBlocks` is too high will in the worst case lead to a spillover too early. Let's not do anything.